### PR TITLE
Add example support to `xpkg build`

### PR DESCRIPTION
**Pull request by tnthornton**
### Description of your changes
With this change, `xpkg build` now includes examples specified in the `/examples` directory (defaulting to the directory located at the root of the repo or can be configured to a separate location).

In addition `xpkg push` was extended to include support for adding annotations to the the xpkg layers.

I have:

- [X] Read and followed Upbound's [contribution process](https://git.io/fj2m9).
- [X] Run `make reviewable` to ensure this PR is ready for review.

##### xpkg build

Also examines contents:

```
$ _output/bin/darwin_arm64/up -f package -e examples
$ ls
provider-jet-aws-960b41c76b04.xpkg
$ tar -xvf provider-jet-aws-960b41c76b04.xpkg
x sha256:506a64c13fb4987d8d936e92a05e0b5053aa8a62624b75be81e29f225a5f9306
x 3f84d5729bdb7d311816026dc36861485e3c06e23753e27cd15121c352ea9507.tar.gz
x d542afccd9fb9aa14947cd5c2f80850f65f0ceba3b2802488fca542a0b410240.tar.gz
x manifest.json
$  cat manifest.json | jq
[
  {
    "Config": "sha256:506a64c13fb4987d8d936e92a05e0b5053aa8a62624b75be81e29f225a5f9306",
    "RepoTags": null,
    "Layers": [
      "3f84d5729bdb7d311816026dc36861485e3c06e23753e27cd15121c352ea9507.tar.gz",
      "d542afccd9fb9aa14947cd5c2f80850f65f0ceba3b2802488fca542a0b410240.tar.gz"
    ]
  }
]
$ tar -xvf 3f84d5729bdb7d311816026dc36861485e3c06e23753e27cd15121c352ea9507.tar.gz
x package.yaml
$ tar -xvf d542afccd9fb9aa14947cd5c2f80850f65f0ceba3b2802488fca542a0b410240.tar.gz
tar: Removing leading '/' from member names
x .up/examples.yaml
```

##### xpkg push
```
$ _output/bin/darwin_arm64/up xpkg push -f provider-jet-aws-960b41c76b04.xpkg index.docker.io/thorntontn/xpkg-test:v0.0.11
$ docker pull thorntontn/xpkg-test:v0.0.11
v0.0.11: Pulling from thorntontn/xpkg-test
3f84d5729bdb: Pull complete
d542afccd9fb: Pull complete
Digest: sha256:960b41c76b04540aaaba84d7a111a1e743c63c4f25b1fd63bd4e8b8858d73bc3
Status: Downloaded newer image for thorntontn/xpkg-test:v0.0.11
docker.io/thorntontn/xpkg-test:v0.0.11
$ docker manifest inspect thorntontn/xpkg-test:v0.0.11
{
	"schemaVersion": 2,
	"mediaType": "application/vnd.docker.distribution.manifest.v2+json",
	"config": {
		"mediaType": "application/vnd.docker.container.image.v1+json",
		"size": 342,
		"digest": "sha256:506a64c13fb4987d8d936e92a05e0b5053aa8a62624b75be81e29f225a5f9306"
	},
	"layers": [
		{
			"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
			"size": 100635,
			"digest": "sha256:3f84d5729bdb7d311816026dc36861485e3c06e23753e27cd15121c352ea9507",
			"annotations": {
				"io.crossplane.xpkg": "base"
			}
		},
		{
			"mediaType": "application/vnd.docker.image.rootfs.diff.tar.gzip",
			"size": 3796,
			"digest": "sha256:d542afccd9fb9aa14947cd5c2f80850f65f0ceba3b2802488fca542a0b410240",
			"annotations": {
				"io.crossplane.xpkg": "upbound"
			}
		}
	]
}
```

### CODE REVIEW

File: cmd/up/xpkg/push.go [hunk 1]
```diff
@@ -75,7 +83,14 @@ func (c *pushCmd) Run() error {
 	if err != nil {
 		return err
 	}
-	if err := remote.Write(tag, img, remote.WithAuthFromKeychain(
+
+	// annotate image layers
+	aimg, err := annotate(img)
+	if err != nil {
+		return err
+	}
```
Comment (hasheddan) [inline cmd/up/xpkg/push.go:91]: Did some additional thinking about handling the fact that the tarball format is not "round-trippable", or as [ggcr says](https://github.com/google/go-containerregistry/tree/main/pkg/v1/tarball#ubuntulatest):

> This makes it difficult to maintain image digests when roundtripping images through the tarball format, so it's not a great format if you care about provenance.

Ideally we wouldn't have to recreate the image in `up xpkg push` when in most cases we just built it with `up xpkg build`, so I was thinking about ways we could put some interesting metadata in here. One option would be to utilize the fact that the `Config` is both [mutable](https://github.com/google/go-containerregistry/blob/9d1ceb8c5d43571c8883eadcd2722226ed124350/pkg/v1/image.go#L38) and round-trippable with the tarball format. It also happens to support labels that follow the same annotation rules as the manifest descriptor labels (ref: [OCI image config spec](https://github.com/opencontainers/image-spec/blob/main/config.md)). I think a fairly reasonable solution would be to use inject something like `io.crossplane.xpkg.base: <layer-digest>` and `io.crossplane.xpkg.upbound: <layer-digest>` into the labels when we build, then check if they are present when we push and convert them to annotations in the form of `io.crossplane.xpkg: base` and `io.crossplane.xpkg: upbound` on the respective layer descriptors in the manifest. What do you think?

**Review by hasheddan — COMMENTED** [review]
Thanks for this @tnthornton! I left some comments with ideas below. I'd be fine with any of the following options for moving forward:
- Keep as-is and I can follow with changes related to the `--controller` work.
- Drop `up xpkg push` changed and just move forward with the `up xpkg build` functionality that you have implemented.
- Keep iterating in this PR to get to our desired end state.

Let me know what your preference is based on bandwidth 👍🏻

File: cmd/up/xpkg/push.go [hunk 2]
```diff
@@ -87,3 +102,54 @@ func (c *pushCmd) Run() error {
 	}
 	return nil
 }
+
+// annotate reads in the layers of the given v1.Image and annotates the xpkg
+// layers with their corresponding annotations, returning a new v1.Image
+// containing the annotation details.
```
[the remark this answers is not in the document]
Comment (tnthornton) [reply cmd/up/xpkg/push.go:108]: Pretty close. The image is assumed to already be in its flattened state. At a high level we:
* Get the image from disk
* Look up the package.yaml
* Lookup the /.up/examples.yaml
* If neither exist, return
* Read in package.yaml, annotate it with "base" annotation.
* Read in /.up/examples.yaml, if it exists, annotate with "upbound" annotation
* Add annotated layers to a new image (sha's are still the same as the contents have not changed)

File: cmd/up/xpkg/build.go [hunk 3]
```diff
@@ -48,25 +51,33 @@ type buildCmd struct {
 
 	Name string `optional:"" help:"Name of the package to be built. Uses name in crossplane.yaml if not specified. Does not correspond to package tag."`
 
-	PackageRoot string   `short:"f" help:"Path to package directory." default:"."`
-	Ignore      []string `help:"Paths, specified relative to --package-root, to exclude from the package."`
+	PackageRoot  string   `short:"f" help:"Path to package directory." default:"."`
+	ExamplesRoot string   `short:"e" help:"Path to package examples directory." default:"./examples"`
+	Ignore       []string `help:"Paths, specified relative to --package-root, to exclude from the package."`
```
[the remark this answers is not in the document]
Comment (tnthornton) [reply cmd/up/xpkg/build.go:56]: We're actually explicitly ignoring the `examples` dir from the packages dir when we construct the FsBackend.

File: internal/xpkg/build.go [hunk 4]
```diff
@@ -82,17 +84,32 @@ func (t *teeReader) Annotate() interface{} {
 }
 
 // Build compiles a Crossplane package from an on-disk package.
-func Build(ctx context.Context, b parser.Backend, p parser.Parser) (v1.Image, runtime.Object, error) { // nolint:gocyclo
-	// Get YAML stream.
-	r, err := b.Init(ctx)
+// TODO(tnthornton) we should work towards cleaning up this API. It feels
+// pretty clunky at the moment.
```
[the remark this answers is not in the document]
Comment (tnthornton) [reply internal/xpkg/build.go:88]: >Or you could even keep as previously constructed and just pass the image that is being appended to and call Build multiple times.

I believe the Parser interface makes this option explicitly difficult.

>Wonder if we could just pass a slice of parser.Backend + parser.Parser pairs?
We can take this path if we feel strongly about it. This still feels a little clunky, but maybe its the lesser evil.

File: cmd/up/xpkg/push.go [hunk 5]
```diff
@@ -87,3 +102,54 @@ func (c *pushCmd) Run() error {
...
+	fs := tarfs.New(tar.NewReader(reader))
+	pkgYaml, err := fs.Open(xpkg.StreamFile)
+	if err != nil {
+		return nil, err
+	}
+
+	pkgBytes, err := ioutil.ReadAll(pkgYaml)
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (tnthornton) [reply cmd/up/xpkg/push.go:120]: Turns out we need to use afero.FileInfo, but easy enough to get one of those.

File: cmd/up/xpkg/build.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hasheddan) [reply cmd/up/xpkg/build.go:56]: ha would be helpful if I looked a little closer -- you're a step ahead! 🙌🏻

[the remark this answers is not in the document]
Comment (tnthornton) [reply cmd/up/xpkg/build.go:56]: Ha, it actual broke some of my tests before I made that change 😅

File: internal/xpkg/build.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tnthornton) [reply internal/xpkg/build.go]: Ended up taking a slightly different path here. I'm still not quite sold on the constructor's signature. But I feel like it cleans up `Build` quite a bit.

File: Makefile [hunk 6]
```diff
@@ -57,7 +57,7 @@ build.bundle.init:
 	@mkdir -p $(abspath $(OUTPUT_DIR)/bundle/up)
 	@mkdir -p $(abspath $(OUTPUT_DIR)/bundle/docker-credential-up)
 
-ifeq ($(OS),linux)
+ifeq ($(HOSTOS),linux)
```
Comment (tnthornton) [inline Makefile:60]: @hasheddan you should double check this is what you want here. It makes `make build` work for me locally.

File: internal/xpkg/build.go [hunk 7]
```diff
@@ -114,35 +151,55 @@ func Build(ctx context.Context, b parser.Backend, p parser.Parser) (v1.Image, ru
...
 	}
-	if err := tw.Close(); err != nil {
-		return nil, nil, errors.Wrap(err, errTarFromStream)
+	layers = append(layers, pkgLayer)
+
+	// examples exist, create the layer
+	if examplesExist {
+		exBuf := new(bytes.Buffer)
+		_, err = b.ep.Parse(ctx, annotatedTeeReadCloser(exReader, exBuf))
+		if err != nil {
```
Comment (hasheddan) [inline internal/xpkg/build.go:174]: nitpick:
```suggestion
		if _, err = b.ep.Parse(ctx, annotatedTeeReadCloser(exReader, exBuf)); err != nil {
```

File: Makefile [hunk 8]
```diff
@@ -57,8 +57,10 @@ build.bundle.init:
 	@mkdir -p $(abspath $(OUTPUT_DIR)/bundle/up)
 	@mkdir -p $(abspath $(OUTPUT_DIR)/bundle/docker-credential-up)
 
-ifeq ($(OS),linux)
+ifeq ($(OS), linux)
+ifneq ($(HOSTOS), darwin)
```
Comment (hasheddan) [inline Makefile:61]: Does `nfpm` have issues on darwin?

**Review by hasheddan — COMMENTED** [review]
@tnthornton a couple of minor comments, but I think this one is pretty close!

File: Makefile (hunk 8 above)
Comment (tnthornton) [reply Makefile:61]: It appears to. These are the errors I get on mac when running `make build` without the above conditional:
```
fork/exec /var/folders/qh/kb2fj2915zn36r801mw29q_00000gn/T/go-build2496340764/b001/exe/addlicense: exec format error
internal/generate.go:22: running "go": exit status 1
fork/exec /var/folders/qh/kb2fj2915zn36r801mw29q_00000gn/T/go-build3317415189/b001/exe/nfpm: exec format error
internal/build/build.go:23: running "go": exit status 1
```

File: internal/xpkg/build.go [hunk 9]
```diff
@@ -114,35 +151,55 @@ func Build(ctx context.Context, b parser.Backend, p parser.Parser) (v1.Image, ru
...
+
 	return img, meta, nil
 }
+
+// SkipContains supplies a FilterFn that skips paths that contain the give pattern.
+func SkipContains(pattern string) parser.FilterFn {
+	return func(path string, info os.FileInfo) (bool, error) {
+		return strings.Contains(path, pattern), nil
+	}
+}
```
[the remark this answers is not in the document]
Comment (tnthornton) [reply internal/xpkg/build.go:204]: Yes and no. Under the hood SkipPath uses `filepath.Match` which doesn't support `**` glob matching for 1 or more directories. So, if we always dealing with examples being at the root, or if we always new what the root directory was `examples\\*` or `/rootdir/examples\\*` would probably work just fine. In its current form though, `**examples\\*` does not currently match `examples` when its a subdirectory.

[the remark this answers is not in the document]
Comment (hasheddan) [reply internal/xpkg/build.go:204]: Yep, that's what I was referring to with nested `examples/` 👍🏻

File: Makefile (hunk 8 above)
Comment (hasheddan) [reply Makefile:61]: Hmm, that is surprising given that those are tracked using go mod 🤔

File: internal/credhelper/credhelper.go [hunk 10]
```diff
@@ -95,6 +110,15 @@ func (h *Helper) List() (map[string]string, error) {
...
 // Get gets credentials for the supplied server.
 func (h *Helper) Get(serverURL string) (string, string, error) {
+	// check if serverURL is an upbound.io URL OR if the serverURL matches
+	// the configured target endpoint.
+	if !strings.Contains(serverURL, defaultEndpoint) && h.endpoint == "" {
+		return "", "", fmt.Errorf(errNotSupported)
+	}
+	if h.endpoint != "" && serverURL != h.endpoint {
+		return "", "", fmt.Errorf(errNotSupported)
+	}
```
[the remark this answers is not in the document]
Comment (tnthornton) [reply internal/credhelper/credhelper.go:120]: >Additionally, the default in most cases for UP_ENDPOINT is in form of https://api.upbound.io and is parsed as a URL. We may want to handle it as a url.URL here and do a comparison on TLD + 1. (i.e. if UP_ENDPOINT=https://api.my.cool.domain.com we would match any server URL that had domain.com). Also open to adjusting how we use UP_ENDPOINT in other places if you feel that would be better for achieving consistency.

Given this, I'm beginning to wonder if using `UP_ENDPOINT` is the correct choice here or if we should just introduce a new env var for overriding the default registry location.

[the remark this answers is not in the document]
Comment (tnthornton) [reply internal/credhelper/credhelper.go]: Per our convo offline, we're going to handle this in a few ways:
1. remove the current changes to credhelper
2. in a separate issue re-design our usage of `UP_ENDPOINT` across the commands and narrow to `UP_DOMAIN`.
3. redo the credhelper approach to take into account `UP_DOMAIN`

File: Makefile (hunk 8 above)
Comment (hasheddan) [reply Makefile:61]: @tnthornton just checking -- have you tried clearing your go build cache and re-running?

**Review by hasheddan — APPROVED** [review]
Thanks @tnthornton! I feel like something funky is happening with the `go generate` stuff, but I'm having trouble identifying when not on a Mac

File: Makefile (hunk 8 above)
Comment (tnthornton) [reply Makefile:61]: Just tried. I get the same error even when starting from a fresh up checkout (no _outputs, .work, .cache).

Comment (hasheddan) [reply Makefile:61]: Strange -- we can follow up and see if we can correct this, what you have should serve as a suitable workaround for the time-being

File: internal/credhelper/credhelper.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (tnthornton) [reply internal/credhelper/credhelper.go]: Created an issue for this https://github.com/upbound/up/issues/178.

**Pull request merged**